Metrics C&U: Introduce batches

kbrock opened the pull request:
Part of #20071 (dependencies listed there)

also merge
- [ ] https://github.com/ManageIQ/manageiq-providers-vmware/pull/834

metrics perf_capture and callback handle multiple targets

# Flow

perf_capture logic lives in MetricsCapture and obtainable via perf_capture_object method
It is ems specific and queues all work for an ems

```
generic worker:
  scheduler

ems_metrics_coordinator worker:
  => perf_capture_timer => perf_capture_queue => perf_capture_targets => perf_capture_target

ems_collector worker:
  => perf_capture_{realtime,historical}
  => perf_capture_callback (for Host nodes)

ems_metrics_processor worker:
  => perf_rollup_range
  => perf_rollup_range to derive cluster metrics
```

Read: The scheduler calls perf_capture_timer (via the queue)
      perf_capture_timer runs on the ems_metrics_coordiantor worker
      perf_capture_timer calls perf_capture_queue which calls ...

# Before

- perf_capture_targets calls perf_capture_target for each object/timeframe
- message queued for each object/timeframe
- host callback also called for each object/timeframe
- Task used for host callback synchronization to then add rollup when all are done.

# After

- perf_capture_targets calls perf_capture_target with a list of targets
- each ems can override behavior to best meet their needs.
- message queued per 250 objects (Vm and Storage)
- message has same ems, timefame, and type (Host, Vm, or Storage)
- message has same cluster (Host) -- no 250 limit
- host callback called for each cluster (with list of host ids)


# Implementation Details

- Still calling Capture#perf_capture on instance of Vm/Host even though passing target_ids. (instance is one of the targets)
- Wanted to support single target and not move too many objects around.
- The instance_id needs to be sent in the target_ids. (instance_id only used to get correct ems/MetricsCapture subclass)

# Performance

|         ms |        bytes |       objects | query |    qry ms |     rows |`comments`
|        ---:|          ---:|           ---:|   ---:|       ---:|      ---:| ---
|   13,536.0 |  68,468,583* |    10,935,195 | 4,062 |  10,199.5 |      820 |`perf_capture-before`
|    5,045.4 |  50,056,647* |     7,306,457 |    56 |   4,784.2 |      820 |`perf_capture-after`
|        56% |              |               |   99% |       53% |          |
|  701,835.6 | 467,283,634* | 1,180,566,804 |50,855 | 298,696.7 |   12,103 |`capture-before`
|  606,988.0 | 727,409,742* | 1,154,128,824 |38,897 | 236,385.4 |    9,692 |`capture-after`
|        14% |              |               |   24% |       21% |      20% | delta

Fryguy commented on app/models/manageiq/providers/base_manager/metrics_capture.rb:
```diff
@@ -175,18 +175,41 @@ def create_rollup_task_for_cluster(ems_cluster, hosts)
+  # number of records to send at a time for a single capture message
+  # @return [Numeric] 0 for no batching.  (defaults to 0)
+  def batch_size
+    @batch_size ||= ::Settings.dig(:ems, "ems_#{ems.provider_name.underscore}", :capture_batch_size).to_i
```
Is there a better way of getting this settings key @agrare - feels like this should be an Ask Don't Tell situation where we ask the ems for it's settings key?

Fryguy replied:
Why have a setting?  Feels like there should be a method in the provider, and if that method chooses to have a  setting, then it can (and then you also don't need any interpolation here).

I generally have an aversion to settings unless there's a good reason a user might change it.

kbrock replied:
When looking through `settings.yml`, this looked to be the suggested way.

I did have `::Settings.ems["ems_#{ems.provider_name.underscore}"]&.capture_batch_size` but some tests complained/blew up. (I assume because the full settings is not available in tests)

- configure feature on/off by ems (via code or settings)
- configure batching value by ems (via constant or settings)

The way I turned this on/off changed quite a bit while developing this.

kbrock replied:
https://github.com/ManageIQ/manageiq-providers-vmware/pull/834

agrare replied:
Yeah it is too bad the setting isn't `Settings.dig(:ems, ems.emstype)` like it is for `ems_refresh`

If we don't already we could add an `ems_settings` method similar to the `worker_settings`

kbrock replied:
@agrare I can change it to that, I just tried to follow what was out there.

agrare replied:
Not necessary to do here, moving to ems_settings can be in its own PR

Fryguy commented on app/models/manageiq/providers/base_manager/metrics_capture.rb:
```diff
@@ -175,18 +175,41 @@ def create_rollup_task_for_cluster(ems_cluster, hosts)
   def perf_capture_queue_targets(targets, interval, start_time:, end_time:, parent: nil)
-    task = create_rollup_task_for_cluster(parent, targets) if parent
+    if batch_size > 1
+      if parent # send all of the hosts in a cluster in one message
+        perf_capture_queue_target(targets, interval, :start_time => start_time, :end_time => end_time, :rollup => true)
+      else # batch sends
+        targets.each_slice(batch_size) do |targets_batch|
+          perf_capture_queue_target(targets_batch, interval, :start_time => start_time, :end_time => end_time)
+        end
+      end
```
Possible refactoring to DRY up

I also added a comment about _why_ we send all of the hosts in a cluster in one message.  The previous comment doesn't describe that and I think it's important to add that even if we don't go with my refactoring below.

```suggestion
      # send all of the hosts in a cluster in one message and ignore batching
      #   hosts will rollup into the cluster, so the entire batch must be together otherwise
      #   we will require a create_rollup_task_for_cluster
      batch_size = targets.size if parent 

      targets.each_slice(batch_size) do |targets_batch|
        perf_capture_queue_target(targets_batch, interval, :start_time => start_time, :end_time => end_time)
      end
```

kbrock replied:
I like the comment. Put something into the commit message but this needs to be in the code.

I did have the `batch_size = targets.size` in the previous rev, but it will create a bunch of extra arrays -  probably negligible. 

I'll make the change but since there are 2 variables in play, I thought it was easier to read with 2 calls to perf_capture_target (with the extra `rollup` parameter)

Fryguy commented on app/models/manageiq/providers/base_manager/metrics_capture.rb:
```diff
@@ -175,18 +175,41 @@ def create_rollup_task_for_cluster(ems_cluster, hosts)
+  end
+
+  # standard one message per object to collect
+  def perf_capture_queue_targets_individ(targets, interval, start_time:, end_time:, task_id: nil)
```
Not sure on the method name here (I'm usually adverse to abbrevs).

kbrock replied:
I put this in to allow sub classing to override this stuff.
Maybe that is no longer necessary

Fryguy commented on app/models/manageiq/providers/base_manager/metrics_capture.rb:
```diff
@@ -199,23 +222,27 @@ def perf_capture_queue_target(target, interval_name, start_time:, end_time:, tas
-    queue_item[:args] = [start_time, end_time] if start_time
+    queue_item[:args] = [start_time, end_time, target_ids, rollup] if start_time || target_ids.present? || rollup
+
+    # GOAL only use put/put_unless_exists and drop put_or_update
```
Is this comment needed?  Seems you met your goal since the method used it put_unless_exists.

kbrock replied:
probably not. Wanted to make it clear that we want to delete everything after this message.

I'll remove

Fryguy commented on app/models/metric/ci_mixin/capture.rb:
```diff
@@ -34,24 +34,26 @@ def perf_capture(interval_name, start_time = nil, end_time = nil)
+        4.hours.ago.utc
       else # "realtime" for Vm or Host
-        [last_perf_capture_on, 4.hours.ago.utc].compact.max
+        [Array(metrics_capture.target).map(&:last_perf_capture_on).compact.min, 4.hours.ago.utc].compact.max
```
Instead of having metrics_capture be the incoming paramater, can the target/targets themselves be the incoming parameter?

Fryguy commented on app/models/metric/ci_mixin/capture.rb:
```diff
@@ -34,24 +34,26 @@ def perf_capture(interval_name, start_time = nil, end_time = nil)
       case interval_name
       when "historical" # Vm or Host
         Metric::Capture.historical_start_time
       when "hourly" # Storage (value is ignored)
-        last_perf_capture_on || 4.hours.ago.utc
+        4.hours.ago.utc
```
I don't understand this change - Why would we collect more than we need?

kbrock replied:
storages do not use this parameter

I didn't want to do any more work than necessary here for an ignored value

Fryguy commented on spec/models/metric/ci_mixin/rollup_spec.rb:
```diff
@@ -117,6 +120,73 @@
+        expect(messages.size).to eq(2)
+        messages.each do |m|
+          expect(m.args).to match_array([nil, nil, a_kind_of(Array), false])
+          expect(m.args[2].size).to eq(3)
```
Super minor, but I'm wondering if the `:capture_batch_size => 3` setting should just be in this one spec.  I don't think it matters for the other 2 specs?  It took me a while to try to figure out why this was broken up into 2 messages, until I realized it's configured all the way up above.

kbrock replied:
I'm keeping it as is.

I tried moving the batch size down to here but it broke the 3 other tests in this group.
The batch size changes the flow and the way hosts, tasks, and vm batching works.

Fryguy commented on spec/support/metric_helper.rb:
```diff
@@ -37,9 +37,9 @@ def queue_timings(items = MiqQueue.where(:method_name => %w[perf_capture_hourly
-          [s_time, e_time]
+      def arg_day_range(start_time, end_time, *args)
+        (start_time.utc..end_time.utc).step_value(1.day).each_cons(2).collect do |s_time, e_time|
+          [s_time, e_time] + args
```
Every caller of this (at least in this PR) seems to pass `nil, false`, unless I'm missing something.  Would it make sense to just default it here if no args are passed?  It's a little confusing because I think it's directly tied to the signature for `perf_capture`, so it's not like we can pass _anything_ in there args.

Fryguy replied:
I guess what I'm saying is maybe the signature here should be something like (unless my assumptions for the signature matching is wrong):

```ruby
def arg_day_range(start_time, end_time, target_ids = nil, rollup = false) # matches the perf_capture signature
```

kbrock replied:
thanks. looks like they are all the same thing.

Hope in the future we will not segment by dates and just merge all these into a single message.

kbrock commented on app/models/metric/ci_mixin/capture.rb:
```diff
@@ -84,7 +86,7 @@ def just_perf_capture(interval_name, start_time, end_time, metrics_capture)
     log_time << ", end_time: [#{end_time}]" unless end_time.nil?
 
-    _log.info("#{log_header} Capture for #{log_target}#{log_time}...")
+    _log.info("#{log_header} Capture for #{log_target}#{log_time}...") ### and log_target friends
```
we are actually capturing for more than just the target, we have potentialy a slew of targets. But I didn't want to fill the logs with that

kbrock commented:
update:
- rename perf_capture_queue_targets_individ to perf_capture_queue_targets_individual
- dry up call to perf_capture_queue_target by using `each_slice` even when not slicing
- remove GOAL comment
- pass targets instead of metrics_capture into fix_capture_start_end
- fixed logging to show multiple targets are being processed in batch capture
- fixed arg_day_range to dry it up
update:
- fixed arg_day_range 2 (ugh: I didn't get all the changes in)

tried but did not update:
- moving stub_settings with capture_batch_size - every test needed the stub
- defaulting start date for storages (since it is ignored)
- changing ems specific capture_batch_size value

kbrock commented:
@miq-bot cross-repo-tests /providers

kbrock commented:
@miq-bot cross-repo-tests manageiq-providers-vmware#834 /providers

kbrock commented:
🍏 in cross repo: https://github.com/ManageIQ/manageiq-cross_repo-tests/pull/715

Also merge https://github.com/ManageIQ/manageiq-providers-vmware/pull/834

Status: I'm still poking at this, but not getting much of a difference

agrare commented on app/models/manageiq/providers/base_manager/metrics_capture.rb:
```diff
@@ -199,23 +236,26 @@ def perf_capture_queue_target(target, interval_name, start_time:, end_time:, tas
       :state       => ['ready', 'dequeue'],
     }
-    queue_item[:args] = [start_time, end_time] if start_time
+    queue_item[:args] = [start_time, end_time, target_ids, rollup] if start_time || target_ids.present? || rollup
```
Not for this PR, but is there any downside to passing all these options all the time even if they are nil?

kbrock replied:
I would have preferred to skip nulls / use default parameters.
It ended up having quite a few variations for these queries.

agrare commented:
Ran a live test on a vcsim with the full evm_server running,

`Metric::Capture.perf_capture_timer` put batches of IDs on the queue:
```
evm: MIQ(MiqQueue.put) Message id: [1384], Zone: [default], Role: [ems_metrics_collector], Server: [], MiqTask id: [], Handler id: [], Ident: [vmware], Target id: [], Instance id: [5], Task id: [], Command: [ManageIQ::Providers::Vmware::InfraManager::Vm.perf_capture_realtime], Timeout: [600], Priority: [100], State: [ready], Deliver On: [], Data: [], Args: [nil, nil, [5, 6, 7, 12, 13, 14, 19, 20, 24, 1, 2, 3, 4, 8, 9, 10, 11, 15, 16, 17, 18, 21, 22, 23], false]
```

I see 2 queued Host perf captures (one for each of my hosts) is this expected @kbrock ?
```
[----] I, [2022-10-31T13:13:09.144393 #74470:97cc]  INFO -- evm: MIQ(MiqQueue.put) Message id: [1382], Zone: [default], Role: [ems_metrics_collector], Server: [], MiqTask id: [], Handler id: [], Ident: [vmware], Target id: [], Instance id: [1], Task id: [], Command: [ManageIQ::Providers::Vmware::InfraManager::HostEsx.perf_capture_realtime], Timeout: [600], Priority: [100], State: [ready], Deliver On: [], Data: [], Args: [nil, nil, nil, true]
[----] I, [2022-10-31T13:13:09.152770 #74470:97cc]  INFO -- evm: MIQ(MiqQueue.put) Message id: [1383], Zone: [default], Role: [ems_metrics_collector], Server: [], MiqTask id: [], Handler id: [], Ident: [vmware], Target id: [], Instance id: [2], Task id: [], Command: [ManageIQ::Providers::Vmware::InfraManager::HostEsx.perf_capture_realtime], Timeout: [600], Priority: [100], State: [ready], Deliver On: [], Data: [], Args: [nil, nil, nil, true]
```

I see the MetricsCollectorWorker fetching all 24 VM targets at once:
```
[----] I, [2022-10-31T14:05:39.944667 #283654:97cc]  INFO -- evm: MIQ(ManageIQ::Providers::Vmware::InfraManager::MetricsCollectorWorker::Runner#get_message_via_drb) Message id: [1967], Zone: [default], Role: [ems_metrics_collector], Server: [], MiqTask id: [], Handler id: [MiqWorker 277], Ident: [vmware], Target id: [], Instance id: [6], Task id: [], Command: [ManageIQ::Providers::Vmware::InfraManager::Vm.perf_capture_realtime], Timeout: [600], Priority: [100], State: [dequeue], Deliver On: [], Data: [], Args: [nil, nil, [6, 7, 1, 2, 3, 4, 12, 8, 9, 10, 11, 13, 20, 14, 19, 5, 15, 16, 17, 18, 24, 21, 22, 23], false], Dequeued in: [10.779552629] seconds
[----] I, [2022-10-31T14:05:39.944874 #283654:97cc]  INFO -- evm: MIQ(MiqQueue#deliver) Message id: [1967], Delivering...
[----] I, [2022-10-31T14:05:39.948640 #283654:97cc]  INFO -- evm: MIQ(ManageIQ::Providers::Vmware::InfraManager::Vm#just_perf_capture) [realtime] Capture for [ManageIQ::Providers::Vmware::InfraManager::Vm], [24 targets], start_time: [2022-10-31 14:05:39 UTC]...
[----] I, [2022-10-31T14:05:40.376284 #283654:97cc]  INFO -- evm: MIQ(ManageIQ::Providers::Vmware::InfraManager::MetricsCapture#perf_build_query_params) Building query parameters...
[----] I, [2022-10-31T14:05:40.376467 #283654:97cc]  INFO -- evm: MIQ(ManageIQ::Providers::Vmware::InfraManager::MetricsCapture#perf_build_query_params) Building query parameters...Complete
[----] I, [2022-10-31T14:05:40.442166 #283654:97cc]  INFO -- evm: MIQ(ManageIQ::Providers::Vmware::InfraManager::Vm#just_perf_capture) [realtime] Capture for [ManageIQ::Providers::Vmware::InfraManager::Vm], [24 targets], start_time: [2022-10-31 14:05:39 UTC]...Complete - Timings: {:capture_state=>0.2784304618835449, :vim_connect=>0.0439453125, :counter_info=>2.6941299438476562e-05, :capture_intervals=>0.10498619079589844, :build_query_params=>0.00023221969604492188, :num_vim_queries=>24, :vim_execute_time=>0.06008601188659668, :perf_processing=>0.003998994827270508, :num_vim_trips=>2, :total_time=>0.4934260845184326}
[----] I, [2022-10-31T14:05:40.442775 #283654:97cc]  INFO -- evm: MIQ(ManageIQ::Providers::Vmware::InfraManager::Vm#perf_process) [realtime] Processing for ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_H0_VM5], id: [6] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_H0_VM6], id: [7] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_H0_VM0], id: [1] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_H0_VM1], id: [2] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_H0_VM2], id: [3] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_H0_VM3], id: [4] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_C0_RP0_VM3], id: [12] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_H0_VM7], id: [8] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_C0_RP0_VM0], id: [9] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_C0_RP0_VM1], id: [10] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_C0_RP0_VM2], id: [11] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_C0_RP0_VM4], id: [13] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_C0_APP0_VM3], id: [20] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_C0_RP0_VM5], id: [14] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_C0_APP0_VM2], id: [19] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_H0_VM4], id: [5] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_C0_RP0_VM6], id: [15] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_C0_RP0_VM7], id: [16] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_C0_APP0_VM0], id: [17] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_C0_APP0_VM1], id: [18] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_C0_APP0_VM7], id: [24] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_C0_APP0_VM4], id: [21] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_C0_APP0_VM5], id: [22] | ManageIQ::Providers::Vmware::InfraManager::Vm name: [DC0_C0_APP0_VM6], id: [23], for range [2022-10-31T14:05:34-04:00 - 2022-10-31T14:05:34-04:00]...
[----] I, [2022-10-31T14:05:40.482822 #283654:97cc]  INFO -- evm: MIQ(ActiveMetrics::ConnectionAdapters::MiqPostgresAdapter#write_rows) [realtime] Processing 24 performance rows...
[----] I, [2022-10-31T14:05:40.487786 #283654:97cc]  INFO -- evm: EMS: [vcsim], id: [2] Saving EMS Inventory...
[----] I, [2022-10-31T14:05:40.499909 #283654:97cc]  INFO -- evm: EMS: [vcsim], id: [2] Saving EMS Inventory...Complete
[----] I, [2022-10-31T14:05:40.499978 #283654:97cc]  INFO -- evm: MIQ(ActiveMetrics::ConnectionAdapters::MiqPostgresAdapter#write_rows) [realtime] Processing 24 performance rows...Complete - Added 24 / Updated 0
```

Looks right to me just one question about the hosts and if they should be collected together

agrare commented:
> I see 2 queued Host perf captures (one for each of my hosts) is this expected @kbrock ?

Discussed offline and this is because I have one host per cluster and we are checking the length of the targets and defaulting back to one-target-per-item if targets.size == 1
